astropy/wcs/src/wcslib_wrap.c

```diff
@@ -1621,14 +1622,36 @@ PyWcsprm_cset(
     const int convert) {
 
   int status = 0;
+  int cmp = 0;
+  double tolerance = 0;
+  int equal;
+
+  // We want to avoid calling wcsset whenever possible as it is not thread-safe,
+  // so we keep track of what the wcsprm was last time this function was called,
+  // using x_prev, and we compare these. If equal, and if PyWcsprm_cset() succeeded
+  // last time it was called, then we can just return 0 as there should be nothing
+  // to do.
+  if(self->x_prev_set) {
+    wcsprm_python2c(&self->x);
+    wcsprm_python2c(&self->x_prev);
+    status = wcscompare(cmp, tolerance, &self->x, &self->x_prev, &equal);
+    wcsprm_c2python(&self->x);
+    wcsprm_c2python(&self->x_prev);
+    if (status == 0 && equal != 0) return 0;
+  }
 
   if (convert) wcsprm_python2c(&self->x);
   status = wcsset(&self->x);
   if (convert) wcsprm_c2python(&self->x);
 
   if (status == 0) {
```

Comment:
Minor issue: From what I can tell, `wcsset` is called and returns successfully even when given incomplete details, i.e., with `naxis = 0`. May be we can add the `status == 0 && naxis > 0` for only storing a fully initialised `wcsprm` struct.